[prathisttam] iP

prathisttam opened the pull request.

prathisttam committed:
Rename chatbot to Simba and implement greeting and exit

Introduce Simba as the chatbot's new identity. Add a basic greeting and
exit message to improve the user experience.

Also include a decorative ASCII logo for Simba in the greeting.

prathisttam committed:
Update .gitignore to ignore .class files

prathisttam committed:
Add Echo game feature and refactor UI handling

The chatbot lacked interactive features beyond displaying messages,
limiting user engagement.

Keeping UI logic inside Simba.java and Echo.java leads to redundancy and
 makes it difficult to maintain consistency in formatting.

Let's introduce an `Echo` class that allows users to play a simple game
where their input is echoed back until they say bye.

Let's also refactor UI handling into a separate `Ui` class to centralize
 all message formatting and prompts, ensuring consistency and easier
maintenance.

prathisttam committed:
Add AllTasksManager and Task Classes to store and retrieve tasks

There is currently no way for users to add and view tasks in the
chatbot.

Let's add an AllTasksManager class to handle task storage and retrieval.
Tasks are stored in an array and can be added and listed sequentially.

Using an array allows simple indexing and ordering of tasks without
requiring external dependencies. Future enhancements may introduce
dynamic data structures for scalability.

prathisttam committed:
Enable marking and unmarking of tasks as completed

Users cannot currently track task completion status. This change allows
tasks to be marked as completed or incomplete, improving task
management.

prathisttam committed:
Add Javadoc comments across the codebase for clarity

thienkimtranhoang commented on src/main/java/AllTasksManager.java:
```diff
@@ -0,0 +1,98 @@
+        userInterface.printDashedLine();
+        for (int i = 0; i < taskListIndex; i++) {
+            Task currentTask = tasksList[i];
+            System.out.println(taskIndex + ".[" + currentTask.getStatusIcon() + "] " + currentTask.getTaskDescription());
```
Perhaps you should break this line into 2 lines to try to keep line length shorter than 110 characters

thienkimtranhoang commented on src/main/java/AllTasksManager.java:
```diff
@@ -0,0 +1,98 @@
+                String[] parts = input.split(" ");
+                if (parts.length > 1) {
+                    int number = Integer.parseInt(parts[1]);
+                    if (number > 0 && number <= taskListIndex) {
```
I think you should refactor your code more to avoid having too many nested if-els

thienkimtranhoang replied:
I think you should use K&R style brackets here to ensure better code readability

thienkimtranhoang commented on src/main/java/AllTasksManager.java:
```diff
@@ -0,0 +1,98 @@
+                }
+            } else {
+                addTask(input);
+            }
```
Perhaps break this into smaller functions, this segment have quite many nested if-else

thienkimtranhoang commented on src/main/java/Echo.java:
```diff
@@ -0,0 +1,29 @@
+                break;
+            }
+
+            // Echo the command
```
I think comments should be indented relative to their position in the code

whalesyong commented on src/main/java/AllTasksManager.java:
```diff
@@ -1,35 +1,58 @@
 import java.util.Scanner;
 
+/**
```
Great practice of Javadoc comments!!

thienkimtranhoang commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,45 @@
+        System.out.println("____________________________________________________________");
+        System.out.println("Hello! I'm Simba");
+        System.out.println("What can I do for you?\n" + logo);
+        System.out.println("____________________________________________________________");
```
Here I believe have indentaton problems, all println should have the same indentation

whalesyong commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,40 @@
+public class Ui {
```
I like the organization of code, keeping UI elements out of the logic to maintain readability.

whalesyong commented on .gitignore:
```diff
@@ -15,3 +15,6 @@ bin/
 
 /text-ui-test/ACTUAL.TXT
 text-ui-test/EXPECTED-UNIX.TXT
+
```
Good practice!

whalesyong commented on src/main/java/Simba.java:
```diff
@@ -7,6 +7,7 @@ public static void main(String[] args) {
         Echo echoGame = new Echo();
+        AllTasksManager allTasksManager = new AllTasksManager();
 
         while (true) {
```
I think a `switch` statement could be used here for better readability and adherence to standards.

whalesyong reviewed (commented):
Overall, good practices shown! I like the organization and commenting. I think the only minor thing is usage of `switch` instead of `else if`, otherwise good job!

thienkimtranhoang commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,45 @@
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showMenu() {
+        System.out.println("\nChoose an option:");
+        System.out.println("1 - Play the game Echo!");
+        System.out.println("2 - Manage Tasks");
+        System.out.println("3 - Exit");
+        System.out.print("Your choice: ");
+    }
```
I like this piece pf code. Really easy to understand.

thienkimtranhoang commented on src/main/java/Echo.java:
```diff
@@ -0,0 +1,29 @@
+            }
+
+            // Echo the command
+            System.out.println("____________________________________________________________");
+            System.out.println(input);
+            System.out.println("____________________________________________________________");
```
Perhaps can use constant to replace the whole string, making the code easier to read.

thienkimtranhoang commented on src/main/java/AllTasksManager.java:
```diff
@@ -0,0 +1,98 @@
+ * Manages a list of tasks, allowing users to add, list, mark and unmark tasks.
+ */
+public class AllTasksManager {
+    private Task[] tasksList = new Task[100];
```
The 100 in new Task[100] should be a named constant, to avoid magic number.

thienkimtranhoang commented on src/main/java/AllTasksManager.java:
```diff
@@ -0,0 +1,98 @@
+        System.out.println("  [" + currentTask.getStatusIcon() + "] " + currentTask.getTaskDescription());
+    }
+
+    /**
+     * Handles user input for adding, listing, marking and unmarking tasks.
+     * User can enter command "bye" to exit
+     */
```
Good use of comment! Help reviewers to understand the purpose of the next segment of code

thienkimtranhoang reviewed (commented):
Overall really good practices of coding standard and great use of comment! Only minor room for improvement is usage of magic number and introduction of more constant.

prathisttam committed:
Add Todo, Event and Deadlines subclasses of Task

Task management currently does not differentiate between different types
of tasks. Additionally, deeply nested conditional logic in
startTaskManagement reduces readability.

Let's introduce three task types:
- ToDo: Simple tasks without a deadline.
- Deadline: Tasks with a specific due date.
- Event: Tasks that occur within a specific time range.

To improve code clarity, extract handleMark and handleUnmark methods
from startTaskManagement, reducing nested conditionals.

Using inheritance improves code organization and allows polymorphic
handling of tasks and enhances usability by providing clear distinctions
between task types.

prathisttam committed:
Refactor task creation: Remove magic numbers and redundant prints

prathisttam committed:
Chatbot error handling: prevent crashes and improve user feedback

Error handling relies solely on printing messages, making it
inconsistent and difficult to manage. For example, invalid commands
(e.g., deadline /by) do not terminate execution properly, leading to
unintended behavior instead of prompting the user to correct the input.

Users need clear guidance when they enter incorrect commands. Unexpected
crashes break the user experience.

Let's add structured exception handling (SimbaException) to prevent
crashes and guide users.

We now throw exceptions when parameters are missing or invalid, ensuring
that incorrect inputs do not proceed further.

prathisttam committed:
Chatbot error handling: add global exception catch for unexpected errors

Currently, only known user input errors (SimbaException) are handled,
but unexpected runtime errors (e.g., NullPointerException) can crash
the program.

Crashes disrupt the user experience and make debugging harder.

Let's add a global exception catch (`Exception e`) to handle all
unexpected errors. This prevents abrupt crashes and provides a clear
error message, improving stability.

prathisttam committed:
Revert "Chatbot error handling: add global exception catch for unexpected errors"

This reverts commit 4c377f99764821da8aea2e4dbdcdbc34d62957a9.

prathisttam committed:
AllTasksManager: Add delete function and refactor to use ArrayList

The task list is implemented using a fixed-size array, which limits the
scalability of task management. Additionally, there is no functionality
to delete tasks, leading to a cumbersome user experience.

Let's refactor the code to use ArrayList<Task> for storing tasks,
providing dynamic sizing and simplifying element manipulation. This
enhances maintainability and scalability.

Additionally, let's add a delete function that leverages the ArrayList's
built-in methods to easily remove tasks by index. This improves the user
experience by allowing efficient task management.

ArrayList is chosen over other collections due to its indexed access and
dynamic resizing, which are ideal for maintaining the order of tasks.

prathisttam committed:
Add storage to save user tasks and reload saved tasks when chatbot
restarts

User data on tasks is lost when chatbot restarts. User has to re-enter
tasks each time they log-in which is impractical.

let's implement automatic saving of tasks whenever the task list changes
and load data from a file when the chatbot starts.

The file will be stored at ./data/duke.txt, and the chatbot will handle
cases where the file or directory does not exist.

Additionally, file paths are specified in a platform-independent way
to ensure compatibility across different operating systems.

prathisttam committed:
Merge branch 'branch-Level-6'

prathisttam committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/simba/AllTasksManager.java

prathisttam committed:
Fix bug caused when merging branches

SaveTasksToStorage function used fixed size array instead of dynamic
array.

Let's update the function to use dynamic array similar to the rest of
the codebase to prevent the program from crashing

okkhoy commented on src/main/java/data/Storage.java:
```diff
@@ -0,0 +1,71 @@
+
+public class Storage {
+
+    private static final String FILE_PATH = "data/Simba.txt";
```
consider explicitly using a relative path (e.g., `./data/Simba.txt`)

okkhoy commented on src/main/java/data/Storage.java:
```diff
@@ -0,0 +1,71 @@
+    private static final String FILE_PATH = "data/Simba.txt";
+
+    /**
+     * Constructor: Check if the file and folder exist, create them if they don't
```
format of the header comment doesn't follow cs2113 coding standards

okkhoy commented on src/main/java/data/Storage.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+//    public void saveTasksToStorage(String textToAppend) throws IOException {
+//        try {
+//            FileWriter fw = new FileWriter(FILE_PATH, true);
+//            fw.write(textToAppend + System.lineSeparator());
+//            //fw.close();
+//        } catch (IOException e) {
+//            System.out.println("Error writing to file: " + e.getMessage());
+//            throw e;
+//        }
+//    }
+//
```
don't retain commented out code.

okkhoy commented on src/main/java/simba/AllTasksManager.java:
```diff
@@ -0,0 +1,403 @@
+     * @return A new Deadline object if the input is valid, otherwise returns null and prints an error message.
+     */
+    private Task createDeadline(String taskDetails) throws SimbaException {
+        String[] parts = taskDetails.split(" /by ", 2);
```
consider moving low-level string operations to a new method to get better SLAP

okkhoy commented on src/main/java/simba/AllTasksManager.java:
```diff
@@ -0,0 +1,403 @@
+        }
+    }
+
+}
```
this class is particularly too long. see if you can extract related data/methods to separate classes

okkhoy commented on src/main/java/simba/Echo.java:
```diff
@@ -0,0 +1,31 @@
+    public void startEcho() {
+        System.out.println("Game Started!");
+        System.out.println("____________________________________________________________");
+        //Scanner to read input
```
redundant comment

okkhoy commented on src/main/java/simba/SimbaException.java:
```diff
@@ -0,0 +1,7 @@
+package simba;
+
+public class SimbaException extends Exception{
+    public SimbaException(String message){
```
spacing between `){`

okkhoy commented on src/main/java/simba/Ui.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Ui {
+    public void showWelcomeMessage() {
+        String logo =
```
cute!! :-)

okkhoy reviewed (commented):
@prathisttam 
some of the header comments don't follow the coding standards. 
you can consider improving SLAP of the code. 
otherwise generally neat code. see detailed comments below

prathisttam committed:
Refactor AllTasksManager UI handling to Ui.java class

prathisttam committed:
Refactor Echo class UI handling to Ui.java class

prathisttam committed:
Bug fix for storage of Deadlines and Events

Deadlines and events were stored in an incorrect format, causing data
retrieval failures on program restart.

Re-formatted deadline and event tasks so they can be saved and reloaded
properly.

prathisttam committed:
Add Parser class to deal with making sense of the user command

prathisttam committed:
Move command processing from AllTasksManager to parser to enhance user
input processing

prathisttam committed:
Add find functionality to give users a way to find a task by searching
for a keyword in the task description

prathisttam committed:
Add missing JavaDoc comments for all classes

prathisttam committed:
Merge pull request #1 from prathisttam/branch-Level-9

Add find functionality to give users a way to find a task by searching

prathisttam committed:
Merge branch 'master' into branch-A-JavaDoc

prathisttam committed:
Merge pull request #2 from prathisttam/branch-A-JavaDoc

Add missing JavaDoc comments for all classes

prathisttam committed:
Fix bug to close scanner upon pressing 3

prathisttam committed:
Refactor the classes into suitable java packages

prathisttam committed:
Add a User Guide to the project

User Guide with instructions, features, and command examples improves
user experience with the product.

prathisttam committed:
Add Javadoc comments to overridden methods

Overriden methods in the inheritted classes modify the behavior of the
parent class, but without Javadoc, it is unclear how they differ.

Let's add JavaDoc comments to overriden methods to provide clarity on
method's modified behaviour compared to its parent class and improve
readability.

prathisttam committed:
Format codebase to follow Java coding standards

prathisttam committed:
Update README with link to jar file release

prathisttam committed:
Update instructions on README

prathisttam committed:
Edit README description
